src/metabase/lib/query.cljc

```diff
@@ -188,3 +188,38 @@
    table-id :- [:or ::lib.schema.id/table :string]]
   (let [metadata-provider (lib.metadata/->metadata-provider original-query)]
    (query metadata-provider (lib.metadata/table-or-card metadata-provider table-id))))
+
+(defn- occurs-in-expression?
+  [expression-clause clause-type expression-body]
+  (or (and (lib.util/clause-of-type? expression-clause clause-type)
+           (= (nth expression-clause 2) expression-body))
+      (and (sequential? expression-clause)
+           (boolean
+            (some #(occurs-in-expression? % clause-type expression-body)
+                  (nnext expression-clause))))))
+
+(defn- occurs-in-stage-clause?
+  "Tests whether predicate `pred` is true for an element of clause `clause` of `query-or-join`.
+  The test is transitive over joins."
+  [query-or-join clause pred]
+  (boolean
+   (some (fn [stage]
+           (or (some pred (clause stage))
+               (some #(occurs-in-stage-clause? % clause pred) (:joins stage))))
+         (:stages query-or-join))))
```

Comment:
Not for this PR, but I don't think this is the first time we've wanted a `lib.walk` or `lib.match` library to capture these recursive explorations. We'll probably end up building something like that for the QP eventually.

Reply:
Let's see a few more examples. I grew vary of walking and matching: I had to replace them by custom code because they were the bottlenecks. (I've seen normalization dominate the time needed for I/O.)